stan/math/rev/core/vari.hpp

```diff
@@ -179,11 +165,375 @@ class vari_value<T, std::enable_if_t<std::is_floating_point<T>::value>>
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+  /**
+   * Construct an dense Eigen variable implementation from a value. The
+   *  adjoint is initialized to zero and if `stacked` is `false` this vari
+   *  will be not be put on the var_stack. Instead it will only be put on
+   *  a stack to keep track of whether the adjoint needs to be set to zero.
+   *
+   * All constructed variables are added to a stack.  Variables
+   *  should be constructed before variables on which they depend
+   *  to insure proper partial derivative propagation.  During
+   *  derivative propagation, the chain() method of each variable
+   *  will be called in the reverse order of construction.
+   *
+   * @tparam S A dense Eigen type that is convertible to `value_type`
+   * @param x Value of the constructed variable.
+   * @param stacked If false will put this this vari on the nochain stack so
+   * that its `chain()` method is not called.
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr>
+  vari_value(S&& x, bool stacked)
+      : val_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        adj_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        val_(eigen_map(val_mem_, x.rows(), x.cols()) = x),
+        adj_(eigen_map(adj_mem_, x.rows(), x.cols()).setZero()) {
+    if (stacked) {
+      ChainableStack::instance_->var_stack_.push_back(this);
+    } else {
+      ChainableStack::instance_->var_nochain_stack_.push_back(this);
+    }
+  }
+
+  /**
+   * Return the number of rows for this class's `val_` member
+   */
+  const Eigen::Index rows() const { return val_.rows(); }
+  /**
+   * Return the number of columns for this class's `val_` member
+   */
+  const Eigen::Index cols() const { return val_.rows(); }
+  /**
+   * Return the size of this class's `val_` member
+   */
+  const Eigen::Index size() const { return val_.size(); }
+
+  virtual void chain() {}
+  /**
+   * Initialize the adjoint for this (dependent) variable to 1.
+   * This operation is applied to the dependent variable before
+   * propagating derivatives, setting the derivative of the
+   * result with respect to itself to be 1.
+   */
+  inline void init_dependent() { adj_.setOnes(); }
+
+  /**
+   * Set the adjoint value of this variable to 0.  This is used to
+   * reset adjoints before propagating derivatives again (for
+   * example in a Jacobian calculation).
+   */
+  inline void set_zero_adjoint() final { adj_.setZero(); }
+
+  /**
+   * Insertion operator for vari. Prints the current value and
+   * the adjoint value.
+   *
+   * @param os [in, out] ostream to modify
+   * @param v [in] vari object to print.
+   *
+   * @return The modified ostream.
+   */
+  friend std::ostream& operator<<(std::ostream& os, const vari_value<T>* v) {
+    return os << "val: \n" << v->val_ << " \nadj: \n" << v->adj_;
```

Comment:
I'm not gonna complain if we take the `vari_value` s in by reference here and in the `double` implementation.

Reply:
You mean using a `const vari_value<T>&` instead of a `const vari_value<T>*`?

Reply:
Yah yah, so the way to print a vari would be:

```
vari a;
std::cout << a << std::endl;
```

instead of:
```
vari a;
std::cout << &a << std::endl;
```

Reply:
Oh, yeah sure!

Reply:
Oh ya know what tho. I think we would normally call this like
```
var a(10);
std::cout << a.vi_ << std::endl;
```

So if we make it a ref then we'd have to do
```
var a(10);
std::cout << *a.vi_ << std::endl;
```